Sources/TurboNavigator.swift

```diff
@@ -1,235 +1,73 @@
+import Foundation
 import SafariServices
 import Turbo
 import UIKit
 import WebKit
 
-/// Handles navigation to new URLs using the following rules:
-/// https://github.com/joemasilotti/TurboNavigator#handled-flows
-public class TurboNavigator {
-    /// Default initializer.
-    /// - Parameters:
-    ///   - delegate: handle custom controller routing
-    ///   - pathConfiguration: assigned to internal `Session` instances for custom configuration
-    ///   - navigationController: optional: override the main navigation stack
-    ///   - modalNavigationController: optional: override the modal navigation stack
-    public init(delegate: TurboNavigationDelegate,
-                pathConfiguration: PathConfiguration? = nil,
-                navigationController: UINavigationController = UINavigationController(),
-                modalNavigationController: UINavigationController = UINavigationController())
-    {
-        self.session = Session(webView: TurboConfig.shared.makeWebView())
-        self.modalSession = Session(webView: TurboConfig.shared.makeWebView())
-        self.delegate = delegate
-        self.navigationController = navigationController
-        self.modalNavigationController = modalNavigationController
+public class TurboNavigator: TurboNavigationHierarchyControllerDelegate {
+    public unowned var delegate: TurboNavigatorDelegate
 
-        session.delegate = self
-        modalSession.delegate = self
-        session.pathConfiguration = pathConfiguration
-        modalSession.pathConfiguration = pathConfiguration
+    public var rootViewController: UINavigationController { hierarchyController.navigationController }
+
+    public var webkitUIDelegate: TurboWKUIController? {
+        didSet {
+            session.webView.uiDelegate = webkitUIDelegate
+            modalSession.webView.uiDelegate = webkitUIDelegate
+        }
```

Comment:
Oh should we make this `public var webkitUIDelegate: WKUIDelegate?` instead here? Otherwise I think we run into the issue @joemasilotti raised [here](https://github.com/joemasilotti/TurboNavigator/pull/72#issuecomment-1789610912) about having to keep our `TurboWKUIController` in step with the WKUIDelegate protocol that we don't control, because subclasses can't implement protocol methods that aren't in the superclass. 

And it'd be nice for users to be able to opt out of the alerting behaviour if they want, by supplying their own WKUIDelegate. 

Perhaps we could set the default TurboWKUIController up as part of the `convenience init` on TurboNavigator, if the intention there is to provide the easiest route for set up 🤔?

Reply:
Great catch! One caveat is that `TurboWKUIController` is a class, not a protocol. So a developer can subclass it to customize alert/confirm behavior or add new functionality.

But doing so is a little awkward because it needs a reference back to `TurboNavigator`:

```swift
self.turboNavigator = TurboNavigator(pathConfiguration: self.pathConfiguration)
self.turboNavigator.webkitUIDelegate = WKUIController(delegate: self.turboNavigator)

class WKUIController: TurboWKUIController {
    // Override functions to customize behavior, presenting alerts on `delegate`.
}
```

If we expose a generic `WKUIDelegate` then I don't see how a developer can grab a reference back to the `TurboWKUIDelegate` needed to actually do the alert presenting!

Here's the diff I'm working with to allow subclassing:

```diff
diff --git a/Sources/TurboWKUIDelegate.swift b/Sources/TurboWKUIDelegate.swift
index 94eb1d4..f14ebba 100644
--- a/Sources/TurboWKUIDelegate.swift
+++ b/Sources/TurboWKUIDelegate.swift
@@ -5,9 +5,10 @@ public protocol TurboWKUIDelegate : AnyObject {
     func present(_ alert: UIAlertController, animated: Bool)
 }

-public class TurboWKUIController : NSObject, WKUIDelegate {
-    weak var delegate: TurboWKUIDelegate?
-    init(delegate: TurboWKUIDelegate!) {
+open class TurboWKUIController: NSObject, WKUIDelegate {
+    public private(set) weak var delegate: TurboWKUIDelegate?
+
+    public init(delegate: TurboWKUIDelegate!) {
         self.delegate = delegate
     }
```

Reply:
Ohh that's a good point about the delegate. Would wrapping like this work do you reckon? So you'd be able to pass in either a subclassed `TurboWKUIController` or your own `WKUIDelegate` and have either be accessible from the outside? 

```swift 
    public enum WKUIDelegateType {
        case standardBehaviour(TurboWKUIController)
        case custom(WKUIDelegate)
        
        var delegate: WKUIDelegate {
            switch self {
            case .custom(let delegate): return delegate
            case .standardBehaviour(let controller): return controller
            }
        }
    }
 ...

    public var webkitUIDelegate: WKUIDelegateType? {
        didSet {
            session.webView.uiDelegate = webkitUIDelegate?.delegate
            modalSession.webView.uiDelegate = webkitUIDelegate?.delegate
        }
    }


... // then in the init method: 

        defer { self.webkitUIDelegate = .standardBehaviour(TurboWKUIController(delegate: self)) }
```

Reply:
> Oh should we make this public var webkitUIDelegate: WKUIDelegate? instead here?

I went with a class implementing `WKUIDelegate` so users could subclass it and override any protocol function in WKUIDelegate. If we explosed `WKUIDelegate` we leave all the responsibility of catching and building alarms to the consumer. We should definitely allow that flexibility, just not as a default. `TurboWKUIController` is the default.

Having said that...

> subclasses can't implement protocol methods that aren't in the superclass.

This is an interesting idea! I see `WKUIDelegate` is an Obj-C protocol, so your observation is great: what happens if `TurboWKUIDelegate` does not conform to an optional method? Can users override it? I actually didn't know if this was true or not so I just tested it in a Playground and I don't think it's entirely true.

```swift
import UIKit

@objc protocol TestProtocol {
    @objc optional func doSomething()
    @objc optional func saySomething()
}

class MyClassA : TestProtocol {
    func doSomething() { 
        print("Do A")
    }
}

class MyClassB : MyClassA {
    override func doSomething() {
        print("Do B")
    }
    
    func saySomething() {
        print("Say")
    }
}

let classA = MyClassA()
let classB = MyClassB()

classA.doSomething()
classA.saySomething?() // ❌ Value of type 'MyClassA' has no member 'saySomething'

classB.saySomething()
classB.doSomething()
``` 

*Note that `MyClassB.doSomething` is overridden but `MyClassB.saySomething` is not.

If you comment out `classA.saySomething?()` the code compiles and behaves as expected. This is good news as it means we don't actually need to implement all methods in `WKUIDelegate`. We can conform to however many we want, and users could just conform to functions that we don't implement.

Reply:
> Ohh that's a good point about the delegate. Would wrapping like this work do you reckon? So you'd be able to pass in either a subclassed TurboWKUIController or your own WKUIDelegate and have either be accessible from the outside?

I don't think I'm understanding the idea here. Why are we adding another layer of indirection via the enum? In my mind, a user would just subclass `TurboWKUIDelegate` and just pass their instance as the controller. As soon as it's passed, it's set as the WKUI delegate.

If the issue is the default, we can just have TurboNavigator build its own `TurboWKUIDelegate`.

Reply:
I just pushed two commits that refactor both delegates a bit. I think this should get us everything we need. Check out the `ComplexSceneDelegate` at the bottom of `SceneDelegate.swift`.

The changes move the default implementation of `TurboNavigatorDelegate` to a protocol extension. Which means developers only have to implement the functions they want.

And makes `TurboWKUIController` open for subclassing, allowing a developer to customize only what they need.

I'm pretty happy with this! And I think `ComplexSceneDelegate` shows how it all works together. Let me know if I missed something or drifted off target.

Reply:
@olivaresf Ohhhh you're totally right, apologies for the detour - I'd got mixed up with another subclassing/protocol scenario which I'll no doubt bore you with over coffee sometime 🙇‍♀️

@joemasilotti The new commits look great – I'm with Fernando, let's start to upstream and get real 😃